README.md
# Star History

[![Star History Chart](https://api.star-history.com/svg?repos=hackware1993/Flutter_ConstraintLayout&type=Date)](https://star-history.com/#hackware1993/Flutter_ConstraintLayout)

# Flutter ConstraintLayout

[简体中文](https://github.com/hackware1993/flutter-constraintlayout/blob/master/README_CN.md)

**I have developed the world's fastest general purpose sorting algorithm, which is on average 3
times faster than Quicksort and up to 20 times faster**
, [ChenSort](https://github.com/hackware1993/ChenSort)

A super powerful Stack, build flexible layouts with constraints. Similar to ConstraintLayout for
Android and AutoLayout for iOS. But the code implementation is much more efficient, it has O(n)
layout time complexity and no linear equation solving is required.

**It surpasses traditional nested writing in terms of performance, flexibility, development speed,
and maintainability. It pretty much negates the O(2n) layout algorithm of Intrinsic Measurement.**

**It is a layout and a more modern general layout framework.**

# Greatly improve Flutter development experience and efficiency. Improve application performance

No matter how complex the layout is and how deep the constraints are, it has almost the same
performance as a single Flex or Stack. When facing complex layouts, it provides better performance,
flexibility, and a very flat code hierarchy than Flex and Stack. Say no to 'nested hell'.

In short, once you use it, you can't go back.

Improving "nested hell" is one of my original intentions for developing Flutter ConstraintLayout,
but I don't advocate the ultimate pursuit of one level of nesting, which is unnecessary. So features
like chains are already well supported by Flex itself, so ConstraintLayout will not actively support
it.

View [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

**Flutter ConstraintLayout has extremely high layout performance. It does not require linear
equations to solve.** At any time, each child element will only be laid out once. When its own width
or height is set to wrapContent, some child elements may calculate the offset twice. The layout
process of ConstraintLayout consists of the following three steps:

1. Constraint calculation
2. Layout
3. Draw

The performance of layout and drawing is almost equivalent to a single Flex or Stack, and the
performance of constraint calculation is roughly 0.01 milliseconds (layout of general complexity, 20
child elements). Constraints are only recalculated after they have changed.

ConstraintLayout itself can be arbitrarily nested without performance issues, each child element in
the render tree is only laid out once, and the time complexity is O(n) instead of O(2n) or worse.

A smaller Widget tree leads to less build time and a smaller Element tree. A very flat layout
structure results in a smaller RenderObject tree and less rendering time. One thing most people tend
to overlook is that complex nesting can cause build times to sometimes exceed render times.

It is recommended to use ConstraintLayout at the top level. For extremely complex layout(one
thousand child elements, two thousand constraints), layout and drawing total time within 5
milliseconds(debug mode on Windows 10，release mode take less time), the frame rate can be easily
reached 200 fps.

**If not necessary, try to be relative to the parent layout, so that you can define less id. Or use
relative id.**

**Warning**:
For layout performance considerations, constraints are always one-way, and there should be no two
child elements directly or indirectly restrain each other(for example, the right side of A is
constrained to the left side of B, and the left side of B is in turn constrained to A right). Each
constraint should describe exactly where the child elements are located. Although constraints can
only be one-way, you can still better handle things that were previously (Android ConstraintLayout)
two-way constraints, such as chains(not yet supported, please use with Flex).

# Feature

1. build flexible layouts with constraints
    1. left
        1. toLeft
        2. toCenter(with bias, the default value is 0.5)
        3. toRight
    2. right
        1. toLeft
        2. toCenter(with bias, the default value is 0.5)
        3. toRight
    3. top
        1. toTop
        2. toCenter(with bias, the default value is 0.5)
        3. toBottom
    4. bottom
        1. toTop
        2. toCenter(with bias, the default value is 0.5)
        3. toBottom
    5. baseline
        1. toTop
        2. toCenter(with bias, the default value is 0.5)
        3. toBaseline
        4. toBottom
2. margin and goneMargin(when the visibility of the dependent element is gone or the actual size of
   one side is 0, the goneMargin will take effect, otherwise the margin will take effect, even if
   its own visibility is gone)
3. clickPadding(quickly expand the click area of child elements without changing their actual size.
   This means that you can completely follow the layout of the UI prototype without having to think
   about the click area of the element. This also means that the click area can be shared between
   child elements without increasing nesting. Sometimes it may be necessary to combine with e-index)
4. visibility control
    1. visible
    2. invisible
    3. gone(sometimes it may be better to use a conditional expression to keep the element from
       being created)
5. constraint integrity hint
6. bias(when there are constraints left and right or top and bottom, horizontalBias and verticalBias
   can be used to adjust the offset. The default value is 0.5, which means centering)
7. z-index(drawing order, default is child index)
8. translate、rotate
9. percentage layout(when size is set to matchConstraint, the percentage layout will take effect,
   the default percentage is 1 (100%). The relevant properties are widthPercent, heightPercent,
   widthPercentageAnchor, heightPercentageAnchor)
10. guideline
11. constraints and widgets separation
12. barrier
13. dimension ratio
    1. widthHeightRatio: 1 / 3,
    2. ratioBaseOnWidth: true, (the default value is null, which means automatic inference. The size
       of the undetermined side will be calculated using the determined side based on the aspect
       ratio. The undetermined side must be matchConstraint, and the determined side can be
       matchParent, fixed size(>=0), matchConstraint)
14. relative id(if an id is defined for a child element, it cannot be referenced using a relative
    id)
    1. rId(3) represents the 3th child element, and so on
    2. rId(-1) represents the last child element
    3. rId(-2) represents the penultimate child element, and so on
    4. sId(-1) represents the previous sibling element, and so on
    5. sId(1) represents the next sibling element, and so on
15. wrapper constraints
    1. topLeftTo
    2. topCenterTo
    3. topRightTo
    4. centerLeftTo
    5. centerTo
    6. centerRightTo
    7. bottomLeftTo
    8. bottomCenterTo
    9. bottomRightTo
    10. centerHorizontalTo
    11. centerVerticalTo
    12. outTopLeftTo
    13. outTopCenterTo
    14. outTopRightTo
    15. outCenterLeftTo
    16. outCenterRightTo
    17. outBottomLeftTo
    18. outBottomCenterTo
    19. outBottomRightTo
    20. centerTopLeftTo
    21. centerTopCenterTo
    22. centerTopRightTo
    23. centerCenterLeftTo
    24. centerCenterRightTo
    25. centerBottomLeftTo
    26. centerBottomCenterTo
    27. centerBottomRightTo
16. staggered grid、grid、list(list is a special staggered grid, grid is also a special staggered
    grid)
17. circle position
18. pinned position
19. arbitrary position
20. e-index(event dispatch order, default is z-index)
21. the size of child widgets can be set to:
    1. fixed size(>=0)
    2. matchParent
    3. wrapContent(default, minimum and maximum supported)
    4. matchConstraint
22. the size of itself can be set to:
    1. fixed size(>=0)
    2. matchParent(default)
    3. wrapContent(minimum and maximum are temporarily not supported)
23. layout debugging
    1. showHelperWidgets
    2. showClickArea
    3. showZIndex
    4. showChildDepth
    5. debugPrintConstraints
    6. showLayoutPerformanceOverlay
24. open grammar gives you more flexibility to organize child elements, solving the problem of using
    only limited expressions in the children list

```dart
class OpenGrammarExample extends StatelessWidget {
  const OpenGrammarExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: ConstraintLayout().open(() {
        if (DateTime
            .now()
            .millisecond % 2 == 0) {
          Container(
            color: Colors.red,
          ).applyConstraint(
            size: 200,
            centerTo: parent,
          );
        } else {
          Container(
            color: Colors.yellow,
          ).applyConstraint(
            size: 200,
            centerTo: parent,
          );
        }

        for (int i = 0; i < 5; i++) {
          Row().open(() {
            for (int j = 0; j < 10; j++) {
              Text("$i x $j").enter();
              const SizedBox(
                width: 20,
              ).enter();
            }
          }).applyConstraint(
            height: 100,
            left: parent.left.margin(100),
            top: i == 0 ? parent.top : sId(-1).bottom,
          );
        }

        int i = 0;
        while (i < 100) {
          Text("$i").applyConstraint(
            left: parent.left,
            top: i == 0 ? parent.top : sId(-1).bottom,
          );
          i++;
        }
      }),
    );
  }
}
```

Follow-up development plan:

1. chain
2. constraints visualization
3. provides a visual editor to create layouts by dragging and dropping
4. automatically convert design drafts into code
5. more…

Subscribe to my WeChat official account to get the latest news of ConstraintLayout([Full Guide](https://mp.weixin.qq.com/s/yTpp7TtLagOeWzX2URKOIQ), Best
Practices, Principle Analysis). Follow-up will also share some high-quality, unique, and thoughtful
Flutter technical articles.
![official_account.webp](https://github.com/hackware1993/flutter-constraintlayout/blob/master/official_account.webp?raw=true)

Currently, I am developing a new declarative UI framework weiV(pronounced the same as wave) for
Android based on the View system. It has the following advantages:

1. Declarative UI writing doubles the efficiency of native development
2. Meets or exceeds the performance of the View system
    1. I ported my Flutter ConstraintLayout to Android, relying on its advanced layout algorithm,
       without introducing Intrinsic Measurement, so that the child elements in the View tree will
       only be laid out once in any case, making arbitrary nesting does not cause performance
       issues. Even if each level in the View tree is a mix of wrap_content and match_parent
    2. xml will be discarded
3. All your existing View system experience will be retained
4. All existing UI components will be reused
5. It is written in Kotlin but supports Java friendly
6. At present, the initial support for real-time effective dynamization has begun. You can issue JS,
   use JS to write page logic, and generate JSON describing the Widget tree and pass it to the
   native, and the native uses a non-reflection method to convert it into a real Widget tree and
   render. I might consider implementing a declarative API in JS later
7. In the future, the same stateful hot reload as Flutter will be implemented

The sample code is as follows:

```kotlin
class WeiVCounterKotlinActivity : WeiVActivity() {
    private var count = 0
    private val maxCount = 10
    private val minCount = 0

    override fun build() = WeiV {
        Flex {
            it.orientation = FlexDirection.VERTICAL

            Button(text = "Add count", enable = count < maxCount, onClick = {
                setState {
                    count++
                }
            })

            Button(text = "Sub count", enable = count > minCount, onClick = {
                setState {
                    count--
                }
            })

            Text(text = "count = $count")
        }
    }
}
```

**No one wants to overturn their past experience with the View system, Compose's design is too
bad.**

[weiV GitHub Home Page](https://github.com/hackware1993/weiV)

Subscribe to my WeChat official account to get the latest news of weiV.

Support platform:

1. Android
2. iOS
3. Mac
4. Windows
5. Linux
6. Web

# Import

Null-safety

```yaml
dependencies:
  flutter_constraintlayout:
    git:
      url: 'https://github.com/hackware1993/Flutter-ConstraintLayout.git'
      ref: 'v1.7.0-stable'
```

```yaml
dependencies:
  flutter_constraintlayout: ^1.7.0-stable
```

```dart
import 'package:flutter_constraintlayout/flutter_constraintlayout.dart';
```

# Example [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

![effect.gif](https://github.com/hackware1993/flutter-constraintlayout/blob/master/effect.gif?raw=true)

```dart
class SummaryExampleState extends State<SummaryExample> {
  double x = 0;
  double y = 0;

  ConstraintId box0 = ConstraintId('box0');
  ConstraintId box1 = ConstraintId('box1');
  ConstraintId box2 = ConstraintId('box2');
  ConstraintId box3 = ConstraintId('box3');
  ConstraintId box4 = ConstraintId('box4');
  ConstraintId box5 = ConstraintId('box5');
  ConstraintId box6 = ConstraintId('box6');
  ConstraintId box7 = ConstraintId('box7');
  ConstraintId box8 = ConstraintId('box8');
  ConstraintId box9 = ConstraintId('box9');
  ConstraintId box10 = ConstraintId('box10');
  ConstraintId box11 = ConstraintId('box11');
  ConstraintId barrier = ConstraintId('barrier');

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: const CustomAppBar(
        title: 'Summary',
        codePath: 'example/summary.dart',
      ),
      backgroundColor: Colors.black,
      body: ConstraintLayout(
        // Constraints can be separated from widgets
        childConstraints: [
          Constraint(
            id: box0,
            size: 200,
            bottomLeftTo: parent,
            zIndex: 20,
          )
        ],
        children: [
          Container(
            color: Colors.redAccent,
            alignment: Alignment.center,
            child: const Text('box0'),
          ).applyConstraintId(
            id: box0, // Constraints can be separated from widgets
          ),
          Container(
            color: Colors.redAccent,
            alignment: Alignment.center,
            child: const Text('box1'),
          ).apply(
            constraint: Constraint(
              // Constraints set with widgets
              id: box1,
              width: 200,
              height: 100,
              topRightTo: parent,
            ),
          ),
          Container(
            color: Colors.blue,
            alignment: Alignment.center,
            child: const Text('box2'),
          ).applyConstraint(
            // Constraints set with widgets easy way
            id: box2,
            size: matchConstraint,
            centerHorizontalTo: box3,
            top: box3.bottom,
            bottom: parent.bottom,
          ),
          Container(
            color: Colors.orange,
            width: 200,
            height: 150,
            alignment: Alignment.center,
            child: const Text('box3'),
          ).applyConstraint(
            id: box3,
            right: box1.left,
            top: box1.bottom,
          ),
          Container(
            color: Colors.redAccent,
            alignment: Alignment.center,
            child: const Text('box4'),
          ).applyConstraint(
            id: box4,
            size: 50,
            bottomRightTo: parent,
          ),
          GestureDetector(
            child: Container(
              color: Colors.pink,
              alignment: Alignment.center,
              child: const Text('box5 draggable'),
            ),
            onPanUpdate: (details) {
              setState(() {
                x += details.delta.dx;
                y += details.delta.dy;
              });
            },
          ).applyConstraint(
            id: box5,
            width: 120,
            height: 100,
            centerTo: parent,
            zIndex: 100,
            translate: Offset(x, y),
            translateConstraint: true,
          ),
          Container(
            color: Colors.lightGreen,
            alignment: Alignment.center,
            child: const Text('box6'),
          ).applyConstraint(
            id: box6,
            size: 120,
            centerVerticalTo: box2,
            verticalBias: 0.8,
            left: box3.right,
            right: parent.right,
          ),
          Container(
            color: Colors.lightGreen,
            alignment: Alignment.center,
            child: const Text('box7'),
          ).applyConstraint(
            id: box7,
            size: matchConstraint,
            left: parent.left,
            right: box3.left,
            centerVerticalTo: parent,
            margin: const EdgeInsets.all(50),
          ),
          Container(
            color: Colors.cyan,
            alignment: Alignment.center,
            child: const Text('child[7] pinned to the top right'),
          ).applyConstraint(
            width: 200,
            height: 100,
            left: box5.right,
            bottom: box5.top,
          ),
          const Text(
            'box9 baseline to box7',
            style: TextStyle(
              color: Colors.white,
            ),
          ).applyConstraint(
            id: box9,
            baseline: box7.baseline,
            left: box7.left,
          ),
          Container(
            color: Colors.yellow,
            alignment: Alignment.bottomCenter,
            child: const Text(
                'percentage layout\nwidth: 50% of parent\nheight: 30% of parent'),
          ).applyConstraint(
            size: matchConstraint,
            widthPercent: 0.5,
            heightPercent: 0.3,
            horizontalBias: 0,
            verticalBias: 0,
            centerTo: parent,
          ),
          Barrier(
            id: barrier,
            direction: BarrierDirection.left,
            referencedIds: [box6, box5],
          ),
          Container(
            color: const Color(0xFFFFD500),
            alignment: Alignment.center,
            child: const Text('align to barrier'),
          ).applyConstraint(
            width: 100,
            height: 200,
            top: box5.top,
            right: barrier.left,
          )
        ],
      ),
    );
  }
}
```

# Advanced usage

1. guideline [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

![guideline.webp](https://github.com/hackware1993/flutter-constraintlayout/blob/master/guideline.webp?raw=true)

```dart
class GuidelineExample extends StatelessWidget {
  const GuidelineExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    ConstraintId guideline = ConstraintId('guideline');
    return MaterialApp(
      home: Scaffold(
        body: ConstraintLayout(
          children: [
            Container(
              color: const Color(0xFF005BBB),
            ).applyConstraint(
              width: matchParent,
              height: matchConstraint,
              top: parent.top,
              bottom: guideline.top,
            ),
            Guideline(
              id: guideline,
              horizontal: true,
              guidelinePercent: 0.5,
            ),
            Container(
              color: const Color(0xFFFFD500),
            ).applyConstraint(
              width: matchParent,
              height: matchConstraint,
              top: guideline.bottom,
              bottom: parent.bottom,
            ),
            const Text(
              'Stand with the people of Ukraine',
              style: TextStyle(
                fontSize: 40,
                color: Colors.white,
              ),
            ).applyConstraint(
              width: wrapContent,
              height: wrapContent,
              centerHorizontalTo: parent,
              bottom: guideline.bottom,
            )
          ],
        ),
      ),
    );
  }
}
```

2. barrier [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

![barrier.gif](https://github.com/hackware1993/flutter-constraintlayout/blob/master/barrier.gif?raw=true)

```dart
class BarrierExample extends StatelessWidget {
  const BarrierExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    ConstraintId leftChild = ConstraintId('leftChild');
    ConstraintId rightChild = ConstraintId('rightChild');
    ConstraintId barrier = ConstraintId('barrier');
    return MaterialApp(
      home: Scaffold(
        body: ConstraintLayout(
          debugShowGuideline: true,
          children: [
            Container(
              color: const Color(0xFF005BBB),
            ).applyConstraint(
              id: leftChild,
              width: 200,
              height: 200,
              top: parent.top,
              left: parent.left,
            ),
            Container(
              color: const Color(0xFFFFD500),
            ).applyConstraint(
              id: rightChild,
              width: 200,
              height: matchConstraint,
              right: parent.right,
              top: parent.top,
              bottom: parent.bottom,
              heightPercent: 0.5,
              verticalBias: 0,
            ),
            Barrier(
              id: barrier,
              direction: BarrierDirection.bottom,
              referencedIds: [leftChild, rightChild],
            ),
            const Text(
              'Align to barrier',
              style: TextStyle(
                fontSize: 40,
                color: Colors.blue,
              ),
            ).applyConstraint(
              width: wrapContent,
              height: wrapContent,
              centerHorizontalTo: parent,
              top: barrier.bottom,
              goneMargin: const EdgeInsets.only(top: 20),
            )
          ],
        ),
      ),
    );
  }
}
```

3. badge [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

![badge.webp](https://github.com/hackware1993/flutter-constraintlayout/blob/master/badge.webp?raw=true)

```dart
class BadgeExample extends StatelessWidget {
  const BadgeExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    ConstraintId anchor = ConstraintId('anchor');
    return Scaffold(
      body: ConstraintLayout(
        children: [
          Container(
            color: Colors.yellow,
          ).applyConstraint(
            width: 200,
            height: 200,
            centerTo: parent,
            id: anchor,
          ),
          Container(
            color: Colors.green,
            child: const Text(
              'Indeterminate badge size',
              style: TextStyle(
                color: Colors.black,
                fontSize: 20,
              ),
            ),
          ).applyConstraint(
            left: anchor.right,
            bottom: anchor.top,
            translate: const Offset(-0.5, 0.5),
            percentageTranslate: true,
          ),
          Container(
            color: Colors.green,
          ).applyConstraint(
            width: 100,
            height: 100,
            left: anchor.right,
            right: anchor.right,
            top: anchor.bottom,
            bottom: anchor.bottom,
          )
        ],
      ),
    );
  }
}
```

4. grid [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

![grid.webp](https://github.com/hackware1993/flutter-constraintlayout/blob/master/grid.webp?raw=true)

```dart
class GridExample extends StatelessWidget {
  const GridExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    List<Color> colors = [
      Colors.redAccent,
      Colors.greenAccent,
      Colors.blueAccent,
      Colors.orangeAccent,
      Colors.yellow,
      Colors.pink,
      Colors.lightBlueAccent
    ];
    return Scaffold(
      body: ConstraintLayout(
        children: [
          ...constraintGrid(
              id: ConstraintId('grid'),
              left: parent.left,
              top: parent.top,
              itemCount: 50,
              columnCount: 8,
              itemWidth: 50,
              itemHeight: 50,
              itemBuilder: (index) {
                return Container(
                  color: colors[index % colors.length],
                );
              },
              itemMarginBuilder: (index) {
                return const EdgeInsets.only(
                  left: 10,
                  top: 10,
                );
              })
        ],
      ),
    );
  }
}
```

5. staggered grid [Flutter Web Online Example](https://constraintlayout.flutterfirst.cn)

![staggered_grid.gif](https://github.com/hackware1993/flutter-constraintlayout/blob/master/staggered_grid.gif?raw=true)

```dart
class StaggeredGridExample extends StatelessWidget {
  const StaggeredGridExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    List<Color> colors = [
      Colors.redAccent,
      Colors.greenAccent,
      Colors.blueAccent,
      Colors.orangeAccent,
      Colors.yellow,
      Colors.pink,
      Colors.lightBlueAccent
    ];
    const double smallestSize = 40;
    const int columnCount = 8;
    Random random = Random();
    return Scaffold(
      body: ConstraintLayout(
        children: [
          TextButton(
            onPressed: () {
              (context as Element).markNeedsBuild();
            },
            child: const Text(
              'Upset',
              style: TextStyle(
                fontSize: 32,
                height: 1.5,
              ),
            ),
          ).applyConstraint(
            left: ConstraintId('horizontalList').right,
            top: ConstraintId('horizontalList').top,
          ),
          ...constraintGrid(
              id: ConstraintId('horizontalList'),
              left: parent.left,
              top: parent.top,
              margin: const EdgeInsets.only(
                left: 100,
              ),
              itemCount: 50,
              columnCount: columnCount,
              itemBuilder: (index) {
                return Container(
                  color: colors[index % colors.length],
                  alignment: Alignment.center,
                  child: Text('$index'),
                );
              },
              itemSizeBuilder: (index) {
                if (index == 0) {
                  return const Size(
                      smallestSize * columnCount + 35, smallestSize);
                }
                if (index == 6) {
                  return const Size(smallestSize * 2 + 5, smallestSize);
                }
                if (index == 7) {
                  return const Size(smallestSize * 6 + 25, smallestSize);
                }
                if (index == 19) {
                  return const Size(smallestSize * 2 + 5, smallestSize);
                }
                if (index == 29) {
                  return const Size(smallestSize * 3 + 10, smallestSize);
                }
                return Size(
                    smallestSize, (2 + random.nextInt(4)) * smallestSize);
              },
              itemSpanBuilder: (index) {
                if (index == 0) {
                  return columnCount;
                }
                if (index == 6) {
                  return 2;
                }
                if (index == 7) {
                  return 6;
                }
                if (index == 19) {
                  return 2;
                }
                if (index == 29) {
                  return 3;
                }
                return 1;
              },

CHANGELOG.md
# v1.7.0-stable

adapt flutter 3.13.

# v1.6.3-stable

migrate example.

# v1.6.2-stable

enhance rId.

# v1.6.1-stable

fix pub analysis error.

# v1.6.0-stable

add open grammar support.

# v1.5.2-stable

remove dead code.

# v1.5.1-stable

add paint callback support.

# v1.5.0-stable

1. refactor debug function.
2. add a lot of comments.
3. fix the click problem caused by pinned translate.
4. optimize constraint version implementation.

# v1.4.4-stable

optimize rotate.

# v1.4.3-stable

fix a pinned position bug.

# v1.4.2-stable

fix a rotate bug.

# v1.4.1-stable

add pinned & circle translate support.

# v1.4.0-stable

1. code refactoring, extracting non-core functions.
2. add rotate support.

# v1.3.0-stable

enhance base constraints.

# v1.2.0-stable

add arbitrary position support.

# v1.1.1-stable

add pinned translate support.

# v1.1.0-stable

fix pinned rotate bug.

# v1.1.0-beta

fully support pinned position.

# v1.1.0-alpha

add pinned position support.

# v1.0.7-stable

optimize key set.

# v1.0.6-stable

optimize child wrapContent.

# v1.0.5-stable

optimize constraint calculation performance.

# v1.0.4-stable

shorten package description.

# v1.0.3-stable

reduced constraint computation time to 0.01 ms for normal complexity layouts(10 child elements), 80%
faster performance.

# v1.0.2-stable

optimize layout performance.

# v1.0.1-stable

optimize self wrapContent calculate.

# v1.0.0-stable

simplified id creation.

# v0.9.33-stable

optimize constraints build.

# v0.9.32-stable

fix a margin bug.

# v0.9.31-stable

enhance margin and goneMargin again.

# v0.9.30-stable

1. enhance margin and goneMargin.
2. enhance size set.

# v0.9.29-stable

enhance grid again.

# v0.9.28-beta3

optimize self wrapContent calculate.

# v0.9.28-beta2

fix a self wrapContent bug.

# v0.9.28-beta

fix self wrapContent bugs.

# v0.9.28-alpha2

fix self wrapContent bugs.

# v0.9.28-alpha

add self wrapContent support.

# v0.9.27-stable

add e-index support.

# v0.9.26-stable

enhance relative id.

# v0.9.25-stable

add circle position support.

# v0.9.24-stable

optimize performance overlay.

# v0.9.23-stable

add staggered grid support.

# v0.9.22-stable

add self size set.

# v0.9.21-stable

1. enhance grid again.
2. print child depth.

# v0.9.20-stable

enhance grid.

# v0.9.19-stable

support grid and list.

# v0.9.18-stable

support virtual helper widgets.

# v0.9.17-stable

optimize constraints calculate performance.

# v0.9.16-stable

fix a percentage layout bug.

# v0.9.15-stable

add more wrapper constraints.

# v0.9.14-stable

add relative id.

# v0.9.13-stable

enhance dimension ratio.

# v0.9.12-stable

optimize constraints calculate.

# v0.9.11-stable

remove cache support.

# v0.9.10-stable

cache sharing.

# v0.9.9-stable

optimize code.

# v0.9.8-stable

enhance constraints cache.

# v0.9.7-stable

fix guideline offset.

# v0.9.6-stable

preprocess constraints for performance improve.

# v0.9.5-stable

fix constraints print bug.

# v0.9.4-stable

optimize code.

# v0.9.3-stable

optimize layout performance.

# v0.9.2-stable

fix a bug.

# v0.9.1-stable

fix a bug.

# v0.9.0-stable

support dimension ratio.

# v.0.8.9-stable

support min、max size set.

# v.0.8.8-stable

make parent final.

# v.0.8.7-stable:

just update readme.

# v0.8.6-stable:

The first version published to pub.dev.

The following functions are supported:

1. build flexible layouts with constraints
    1. leftToLeft
    2. leftToRight
    3. rightToLeft
    4. rightToRight
    5. topToTop
    6. topToBottom
    7. bottomToTop
    8. bottomToBottom
    9. baselineToTop
    10. baselineToBottom
    11. baselineToBaseline
2. margin and goneMargin
3. clickPadding (quickly expand the click area of child elements without changing their actual size.
   This means that the click area can be shared between child elements without increasing nesting.
   Sometimes it may be necessary to combine with z-index)
4. visibility control
5. constraint integrity hint
6. bias
7. z-index
8. translate
9. percentage layout
10. guideline
11. constraints and widgets separation
12. barrier

example/home.dart
import 'package:flutter/material.dart';
import 'package:flutter_constraintlayout/flutter_constraintlayout.dart';

import 'arbitrary_position.dart';
import 'badge.dart';
import 'barrier.dart';
import 'charts.dart';
import 'chen_sort.dart';
import 'circle_position.dart';
import 'coming_soon.dart';
import 'complex_list.dart';
import 'constraint_controller.dart';
import 'dimension_ratio.dart';
import 'grid.dart';
import 'guideline.dart';
import 'horizontal_list.dart';
import 'margin.dart';
import 'open_grammar.dart';
import 'percentage_layout.dart';
import 'pinned_position.dart';
import 'relative_id.dart';
import 'self_wrap_content.dart';
import 'staggered_grid.dart';
import 'summary.dart';
import 'translate.dart';
import 'vertical_list.dart';
import 'wrapper_constraints.dart';

class ExampleHome extends StatelessWidget {
  ExampleHome({Key? key}) : super(key: key);

  final Map<String, Widget?> exampleMap = {
    'Chen sort(The fastest sorting algorithm in the world, 60% faster than Quicksort)':
        const ChenSortExample(),
    'Summary': const SummaryExample(),
    'Open Grammar': const OpenGrammarExample(),
    'Guideline': const GuidelineExample(),
    'Barrier': const BarrierExample(),
    'Complex List': const ComplexListExample(),
    'Badge': const BadgeExample(),
    'Percentage Layout': const PercentageLayoutExample(),
    'Dimension Ratio': const DimensionRatioExample(),
    'Relative Id': const RelativeIdExample(),
    'Wrapper Constraints': const WrapperConstraintsExample(),
    'Grid': const GridExample(),
    'Horizontal List': const HorizontalListExample(),
    'Vertical List': const VerticalListExample(),
    'Staggered Grid': const StaggeredGridExample(),
    'Circle Position': const CirclePositionExample(),
    'Pinned Position': const PinnedPositionExample(),
    'Arbitrary Position': const ArbitraryPositionExample(),
    'Constraint Controller': const ConstraintControllerExample(),
    'Self wrapContent': const SelfWrapContentExample(),
    'Margin': const MarginExample(),
    'Translate': const TranslateExample(),
    'Charts': const ChartsExample(),
    'Chain (Coming soon)': const ComingSoonWidget(),
  };

  @override
  Widget build(BuildContext context) {
    List<String> keyList = exampleMap.keys.toList();
    return Scaffold(
      body: ConstraintLayout(
        children: [
          const Text(
            'Flutter ConstraintLayout Example\nby hackware',
            style: TextStyle(
              fontSize: 32,
              height: 1.5,
            ),
            textAlign: TextAlign.center,
          ).applyConstraint(
            topCenterTo: parent,
          ),
          ListView.builder(
            itemBuilder: (_, index) {
              return TextButton(
                onPressed: () {
                  Navigator.of(context).push(MaterialPageRoute(builder: (_) {
                    return exampleMap[keyList[index]]!;
                  }));
                },
                child: Text(
                  keyList[index],
                  style: const TextStyle(
                    fontSize: 20,
                  ),
                ),
              );
            },
            itemCount: keyList.length,
            itemExtent: 40,
          ).applyConstraint(
            width: matchParent,
            height: matchConstraint,
            top: sId(-1).bottom,
            bottom: sId(1).top,
          ),
          const Text(
            'Powered by Flutter Web & ConstraintLayout',
            style: TextStyle(
              fontSize: 16,
              fontWeight: FontWeight.bold,
              height: 1.5,
            ),
          ).applyConstraint(
            bottomCenterTo: parent,
            margin: const EdgeInsets.only(
              bottom: 20,
            ),
          )
        ],
      ),
    );
  }
}
